### ISSUE TITLE
Research Area: L1 Fee Abstraction

### ISSUE THREAD

**Issue by tynes** [ethereum-optimism/specs#73]
Right now, the L1 portion of the fee is defined in native code. With various hardforks, the fee calculations are changed. This adds tech debt to the state transition function as there becomes more branching logic depending on the currently active chain spec. For example, ecotone introduces a new fee formula and fjord may introduce a new fee formula. This means that every op stack execution client needs to support this branching logic.

It is possible to use EVM bytecode instead to define the L1 portion of the fee. A solidity interface could be defined that accepts the RLP encoded transaction and potentially some other data and it is expected to return a `uint256` that represents the L1 fee or the L1 gas used. This would remove altering the fee logic from being a hardfork and increase the social scalability of fee logic discussions as any team working on the OP Stack can choose to use their own fee formulas that work best for them. There should be some standard EVM bytecode that comes with the OP Stack that can do a good job of this fee charging so that teams don't need to think about this problem unless they really want to.

There are some performance considerations of this approach. Benchmarking is required to learn how much of an impact this would have on syncing. Spinning up an EVM instance to do this calculation will always be slower than running the calculation in native code. It is likely that it should be a `STATICCALL` context, ie no writing of state because that could result in some strange side effects. The only application of writing state would be to write the L1 fee to transient storage such that smart contracts would be able to access the value only during the execution of the particular transaction, this could be useful for meta tx relaying.

Another concern is how the L1 portion of the fee isn't committed to by the user's signature. There may have been some work done to attempt to standardize a "multidimensional fee transaction for L2s" that could help with this problem. If the smart contract can be swapped very quickly, then it may be possible for the network to overcharge end users for their L1 fee because they cannot express a "max amount of L1 fee" they are willing to pay. It is possible that this logic could exist in the smart contract doing the L1 fee charging itself, this could work well with smart contract wallets. This is already an issue for what its worth, as the chain operator can bump the L1 fee params in the `SystemConfig` arbitrarily.

**Comment by tynes** [comment]
This could be extended to support more than L1 fee abstraction but instead abstract the full fee payment. This would be more intrusive to EVM equivalence, but would enable a lot of experimentation. It is possible that only abstracting the L1 portion of the fee is enough. We would want to think through the exact data passed into the smart contract if we were abstracting the entire fee payment. Likely we would want `gasUsed`, `success` and the raw transaction data.

**Comment by tynes** [comment]
This could also be used for MEV rebates by reducing fees for users that need a rebate. A mechanism for tracking who deserves rebates would need to be developed. This could be managed by the chain operator as an oracle, or on chain applications could track this information. For example Uniswap v4 hooks could be very useful for automating the process of determining who deserves MEV rebates

**Comment by RealiCZ** [comment]
I believe this proposal has great potential and is also highly flexible. I have carefully reviewed the document. However, there are some critical details that I have doubts about. I hope we can clarify these through discussion. We would like to contribute to accelerating its implementation if necessary.

Fee ratio: The exchange rate between the native token and ether on a custom gas token chain.

> A solidity interface could be defined that accepts the RLP encoded transaction and potentially some other data and it is expected to return a uint256 that represents the L1 fee or the L1 gas used

For this implementation, is a fee ratio needed? If so, where should it be stored, and who should update the fee ratio as it changes frequently?
If not, how should the exchange rate between ether and the native token be handled?

> This could be extended to support more than L1 fee abstraction but instead abstract the full fee payment. 

Since the L1 fee is used to purchase data availability for Ethereum and involves only the data size of the transaction content, it can be computed using a smart contract. However, the L2 fee involves the EVM execution process; how can this approach support it?

**Comment by tynes** [comment]
Regarding the fee ratio, there is no enshrined concept of a fee ratio. It is up to the chain governor to choose the implementation details of the contract that computes the fee. If the chain governor wants to use a fee ratio in their contract, they are free to do so, they could also using an oracle based an an AMM or ecdsa signature.

> For this implementation, is a fee ratio needed? If so, where should it be stored, and who should update the fee ratio as it changes frequently?
If not, how should the exchange rate between ether and the native token be handled?

These are implementation specific details. I am happy to provide advice on how it can be done, but ultimately all the protocol is aware of is EVM execution where bytes are passed in and the amount of gas to be charged is returned. For illiquid assets, you likely want a more centralized oracle solution that is able to push exchange rate updates directly to the L2. For liquid assets, you could use an AMM based oracle for the exchange rate.

> This could be extended to support more than L1 fee abstraction but instead abstract the full fee payment.

Perhaps its best to just keep it simple and only add this sort of abstraction to the L1 portion of the fee. This would be a much more simple change. There wouldn't be a way to turn off the execution fee or modify it but I suppose that is fine.

**Comment by ArtificialPB** [comment]
How would this change impact the `Receipt` type? It currently contains multiple L1 fee-related fields (`L1GasPrice`, `L1BlobBaseFee`, `L1Fee`, ...), some of which were stored in the DB pre-bedrock, but are now decoded from the `L1BlockInfo` tx in `func (rs Receipts) DeriveFields` function. 

The L1 gas price info returned from the on-chain function would then again need to be stored in DB for each `Receipt`, otherwise:
- the receipt lookup (e.g. via RPC) will not return the correct L1 costs associated with the tx execution,
- or, in case the gas cost function would be re-executed on lookup, one would need to run an archive node to be able to compute it.

Do you see re-adding some L1 gas cost fields to `Receipt` as an acceptable outcome of fee abstraction change? Could be implemented similar to `DepositNonce`/`DepositReceiptVersion`.

**Comment by tynes** [comment]
@ArtificialPB I haven't thought super deeply about all of the database implications but you raise good points. You don't want to require an archive node to be able to dynamically recreate the values. It does make sense to index them with the receipt.

**Comment by zobront** [comment]
@tynes Using the example where we have a highly liquid asset and want to pull prices from an AMM, presumably it'd be helpful to have a TWAP so it couldn't be manipulated. Unless that's implemented natively by the AMM, it means that there needs to be some consistent process for caching AMM values to use in those calculations.

I see two possible ways of doing this:

1) Add a preblock action (like the 4788 beacon root call) to get the price from the AMM and store it. This seems pretty non-uniform and prone to error.

2) Change this from a STATICCALL to a CALL, which would allow us to grab the AMM price and save it to our contract's storage before the first transaction in the block from within the call.

The second option seems pretty clearly superior, but curious if you have any specific side effects in mind that pushed you towards wanting to do this as a STATICCALL.

**Comment by tynes** [comment]
The staticcall can be less of a dos vector and I didn't have a good reason to allow for state modifications. Open to the idea but need benchmarking to be sure it's safe

**Comment by zobront** [comment]
@tynes Yeah. Chain operators would definitely need to be careful to be sure whatever functionality they include doesn't impose any DOS risks. 

There's some question about how much to restrict them (ie only allow a fixed amount of gas to be passed to the CALL, with some default to handle failure) vs writing some benchmark tests and trusting them to make the right choices.

Do you have a sense of how OP likes to approach these things? The former probably seems safer to me.

**Comment by tynes** [comment]
We should figure out how much gas is required to be useful for chain operators given a few standard use cases and then benchmark it to see how it impacts disk i/o. Another good benchmark would be to implement the existing fee algo in solidity and see how it impacts sync time on OP Mainnet and Base. There needs to be a cap to the amount of gas passed into the call no matter what but unclear what it should be without understanding how it will be used

**Comment by zobront** [comment]
Reflecting on this more, I don't think my idea will work. We'd need to set a gas cap that is capable of performing the upgrades, but this will be called every transaction (not just every block). That gas cap either needs to be too low to do anything useful, or poses a DOS risk if it's used in full every transaction. Tracking across transactions seems unnecessarily complex.

This brings me back to the other possibility discussed above, where a `CALL` to update values happens independently of the `STATICCALL` to perform the calculation. 

My thought on the simplest option would be that we can piggyback this off the `L1Block.setL1BlockValues()` that is already happening regularly. This could either be accomplished by (a) having `L1Block.setL1BlockValues()` call out to a new Predeploy `L1DataCostOracle.update()` function or (b) by incorporating this functionality directly into the `L1Block` contract.

***

I'd lean towards the second option, as it seems simplest and cleanest. 

On the vanilla OP Stack chain, It would involve:
- Replacing the L1 Data Cost op-geth calculation with a call to `L1Block.getL1DataCost(bytes calldata cd)`.
- Implementing this function on the L1Block contract that mirrors the current logic.

On a CGT chain that uses an AMM for prices, it would also require:
- Adding logic to the `setL1BlockValues()` to read the AMM and store the values locally.
- Writing a new `getL1DataCost()` function that uses those stored values to calculate the correct price.

@tynes Thoughts?
